### PR TITLE
[CI] Update GCC in the dist-linux builders

### PR DESCRIPTION
**Pull request by cuviper**
Upstream LLVM is planning to raise their minimum toolchain requirements, so they may start using C++14 features. This new policy has already landed in the form of a "soft" error.
http://lists.llvm.org/pipermail/llvm-dev/2019-February/130117.html

For GCC, they will require at least version 5.1. This PR moves our crosstool-ng builders to their max GCC 5.2, with a few small patches to fix compatibility. The dist-x86_64-linux builder is updated to GCC 5.5 and LLVM/Clang 8.0.0-rc2, which also affects dist-i686-linux sharing the same scripts.

r? @alexcrichton

### CODE REVIEW

File: src/ci/docker/README.md [hunk 1]
```diff
@@ -131,36 +131,42 @@ $category > $option = $value -- $comment
...
+- Path and misc options > Patches origin = Bundled, then local
+- Path and misc options > Local patch directory = /tmp/patches
 - Target options > Target Architecture = arm
 - Target options > Architecture level = armv6 -- (+)
 - Target options > Floating point = software (no FPU) -- (\*)
 - Operating System > Target OS = linux
 - Operating System > Linux kernel version = 3.2.72 -- Precise kernel
-- C-library > glibc version = 2.14.1
-- C compiler > gcc version = 4.9.3
+- C-library > glibc version = 2.16.0
```
Comment (cuviper) [inline src/ci/docker/README.md]: Note, the arm builders were already configured for glibc 2.16.0, but the README wasn't in sync.

**Comment by alexcrichton** [comment]
@bors: r+

Looks great to me, thanks so much for taking care of this @cuviper!

**Comment by Centril** [comment]
In-between-rollups-filler, @bors p=1

**Comment by kennytm** [comment]
@bors p=187

rollup fairness

**Comment by pietroalbini** [comment]
Travis fixed the issue :tada: 

@bors treeclosed-

**Comment by pietroalbini** [comment]
@bors retry

Spurious network failure.

**Comment by kennytm** [comment]
@bors retry

Normal 3 hour timeout for rebuilding the Docker image.

**Comment by kennytm** [comment]
@bors retry 

3 hour timeout

**Comment by cuviper** [comment]
Hmm, the dist-x86_64 and dist-i686 docker images share most contents locally, only differing in some final environment variables, but I guess our docker caching only operates on the final result? Anyway, it looks like dist-i686 did finish and upload its image before the timeout this time, and everything else passed, so hopefully one more retry will do it...

@bors retry

**Comment by kennytm** [comment]
@bors p=19 

Let's get this merged as soon as possible to avoid invalidating more docker image cache.

**Comment by cuviper** [comment]
Any idea what's invalidating it? It looks like dist-i686-linux has gotten through the docker image a few times, but keeps rebuilding it. The others appear to be cached fine.

**Pull request merged**